client/network/sync/src/engine.rs

```diff
@@ -524,13 +619,194 @@ where
+					hash,
+					who,
+				);
+				return
+			},
+		}
+
+		// Let external validator check the block announcement.
+		let assoc_data = announce.data.as_ref().map_or(&[][..], |v| v.as_slice());
+		let future = self.block_announce_validator.validate(header, assoc_data);
+
+		self.block_announce_validation.push(
+			async move {
+				match future.await {
+					Ok(Validation::Success { is_new_best }) => PreValidateBlockAnnounce::Process {
+						is_new_best: is_new_best || is_best,
+						announce,
+						who,
+					},
+					Ok(Validation::Failure { disconnect }) => {
+						debug!(
+							target: "sync",
+							"Block announcement validation of block {:?} from {} failed",
+							hash,
+							who,
+						);
+						PreValidateBlockAnnounce::Failure { who, disconnect }
+					},
+					Err(e) => {
+						debug!(
+							target: "sync",
+							"💔 Block announcement validation of block {:?} errored: {}",
+							hash,
+							e,
+						);
+						PreValidateBlockAnnounce::Error { who }
+					},
+				}
+			}
+			.boxed(),
+		);
+	}
+
+	/// Poll for finished block announce validations and notify `ChainSync`.
+	fn poll_block_announce_validation(
+		&mut self,
+		cx: &mut std::task::Context,
+	) -> Poll<PollBlockAnnounceValidation<B::Header>> {
+		match self.block_announce_validation.poll_next_unpin(cx) {
+			Poll::Ready(Some(pre_validation)) => {
+				self.deallocate_slot_for_block_announce_validation(pre_validation.who());
+
+				let res = match pre_validation {
+					PreValidateBlockAnnounce::Process { is_new_best, who, announce } => {
+						trace!(
+							target: "sync",
+							"Finished block announce validation: from {:?}: {:?}. local_best={}",
+							who,
+							announce.summary(),
+							is_new_best,
+						);
+						self.chain_sync.on_validated_block_announce(is_new_best, who, &announce);
+						PollBlockAnnounceValidation::Nothing { is_new_best, who, announce }
+					},
+					PreValidateBlockAnnounce::Failure { who, disconnect } => {
+						debug!(
+							target: "sync",
+							"Failed announce validation: {:?}, disconnect: {}",
+							who,
+							disconnect,
+						);
+						PollBlockAnnounceValidation::Failure { who, disconnect }
+					},
+					PreValidateBlockAnnounce::Error { who } => {
+						debug!(
+							target: "sync",
+							"Ignored announce validation from {:?} due to internal validation error",
+							who,
+						);
+						PollBlockAnnounceValidation::Skip
+					},
+				};
+
+				Poll::Ready(res)
+			},
+			_ => Poll::Pending,
+		}
+	}
+
+	/// Checks if there is a slot for a block announce validation.
+	///
+	/// The total number and the number per peer of concurrent block announce validations
+	/// is capped.
+	///
+	/// Returns [`AllocateSlotForBlockAnnounceValidation`] to inform about the result.
+	///
+	/// # Note
+	///
+	/// It is *required* to call [`Self::deallocate_slot_for_block_announce_validation`] when the
+	/// validation is finished to clear the slot.
+	fn allocate_slot_for_block_announce_validation(
```

Comment:
This function could also use some testing

Reply:
Done